kernel/sched.c

```diff
@@ -1095,23 +1106,65 @@ static inline void z_vrfy_k_thread_priority_set(k_tid_t thread, int prio)
 void z_impl_k_thread_deadline_set(k_tid_t tid, int deadline)
 {
 	struct k_thread *thread = tid;
-	int32_t newdl = k_cycle_get_32() + deadline;
+	uint32_t new_dl = k_cycle_get_32() + deadline;
+	k_spinlock_key_t  key;
 
 	/* The prio_deadline field changes the sorting order, so can't
 	 * change it while the thread is in the run queue (dlists
 	 * actually are benign as long as we requeue it before we
 	 * release the lock, but an rbtree will blow up if we break
 	 * sorting!)
 	 */
-	K_SPINLOCK(&_sched_spinlock) {
-		if (z_is_thread_queued(thread)) {
+	key = k_spin_lock(&_sched_spinlock);
+
+	if (z_is_thread_ready(thread)) {
+#if defined(CONFIG_SMP)
```

Comment:
FWIW I think this change is wrong.   There's no valid usage of EDF scheduling that requires setting a deadline on a running thread.  Deadlines get precomputed in all idioms I'm aware of.  If the app needs it to be a scheduling point the app can synthesize one on its own.

Reply:
I did this purely for completeness from a theory POV. However, I must admit lack of knowledge when it comes to practical usage. If you are certain that there is no valid usage EDF scheduling that requires setting a deadline on a running thread, I am happy to defer to your judgement and undo these EDF changes.